pkgs/unified_analytics/lib/src/asserts.dart

```diff
@@ -0,0 +1,126 @@
+    if ((eventMap['name'] as String).length > 40) {
+      throw AnalyticsException('Limit event names to 40 chars or less');
+    }
+    if (!RegExp(r'^[A-Za-z0-9_]+$').hasMatch(eventMap['name'] as String)) {
+      throw AnalyticsException(
+          'Event name can only have alphanumeric chars and underscores');
+    }
+    if (!RegExp(r'^[A-Za-z]+$').hasMatch((eventMap['name'] as String)[0])) {
+      throw AnalyticsException('Event name first char must be alphabetic char');
+    }
+
+    // GA4 Limitation:
+    // Events can have a maximum of 25 parameters
+    if ((eventMap['params'] as Map<String, Object?>).length > 25) {
+      throw AnalyticsException('Limit params for each event to less than 25');
+    }
+
+    // Loop through each of the event parameters
+    (eventMap['params'] as Map<String, Object?>).forEach((key, value) {
+      // GA4 Limitation:
+      // Ensure that each value for the event params is one
+      // of the following types:
+      // `String`, `int`, `double`, or `bool`
+      if (!(value is String ||
+          value is int ||
+          value is double ||
+          value is bool)) {
+        throw AnalyticsException(
+            'Values for event params have to be String, int, double, or bool');
+      }
+
+      // GA4 Limitation:
+      // Parameter names (including item parameters) must be 40 characters
+      // or fewer, may only contain alpha-numeric characters and underscores,
+      // and must start with an alphabetic character
+      if (key.length > 40) {
+        throw AnalyticsException('Limit event param names to 40 chars or less');
+      }
+      if (!RegExp(r'^[A-Za-z0-9_]+$').hasMatch(key)) {
+        throw AnalyticsException(
+            'Event param name can only have alphanumeric chars and underscores');
+      }
+      if (!RegExp(r'^[A-Za-z]+$').hasMatch(key[0])) {
+        throw AnalyticsException(
+            'Event param name first char must be alphabetic char');
+      }
+
+      // GA4 Limitation:
+      // Parameter values (including item parameter values) must be 100
+      // characters or fewer
+      if (value.runtimeType == String) {
+        value as String;
+        if (value.length > 100) {
+          throw AnalyticsException(
+              'Limit characters in event param value to 100 chars or less');
+        }
+      }
+    });
+  }
+
+  // GA4 Limitation:
+  // Events can have a maximum of 25 user properties
+  if (userProperties.length > 25) {
+    throw AnalyticsException('Limit user properties to 25 or less');
+  }
+
+  // Checks for each user property item
+  userProperties.forEach((key, value) {
+    value as Map<String, Object?>;
+
+    // GA4 Limitation:
+    // User property names must be 24 characters or fewer
+    if (key.length > 24) {
+      throw AnalyticsException('Limit user property names to 24 chars or less');
+    }
+
+    // GA4 Limitation:
+    // User property values must be 36 characters or fewer
+    if (value['value'].runtimeType == String) {
```

Comment:
Consider using `is String` rather than checking the runtime type. The VM is allowed to have subclasses of `String` for efficiency, so this test might not always do what you intend. With very few exceptions you should never test the `runtimeType` of an object.

And if you first extract `value['value']` into a local variable, type promotion will make the type cast on the next line unnecessary.

Reply:
Good catch here as well! simplified into one if statement as well